latch_cli/docker_utils/__init__.py

```diff
@@ -0,0 +1,233 @@
+from enum import Enum, auto
+from io import TextIOWrapper
+from pathlib import Path
+from textwrap import dedent
+from typing import List
+
+import yaml
+
+from latch_cli.constants import latch_constants
+from latch_cli.types import LatchWorkflowConfig
+
+
+class DockerCmdBlockOrder(str, Enum):
+    """Put a command block before or after the primary COPY command."""
+
+    precopy = auto()
+    postcopy = auto()
+
+
+@dataclass(frozen=True)
+class DockerCmdBlock:
+    comment: str
+    commands: List[str]
+    order: DockerCmdBlockOrder
+
+    def write_block(self, f: TextIOWrapper):
+        f.write(f"# {self.comment}\n")
+        f.write("\n".join(self.commands) + "\n\n")
+
+
+def get_prologue(config: LatchWorkflowConfig) -> List[str]:
+    return [
+        "# latch base image + dependencies for latch SDK --- removing these will break the workflow",
+        f"from {config.base_image}",
+        f"run pip install latch=={config.latch_version}",
+        f"run mkdir /opt/latch",
+    ]
+
+
+def get_epilogue() -> List[str]:
+    return [
+        "# latch internal tagging system + expected root directory --- changing these lines will break the workflow",
+        "arg tag",
+        "env FLYTE_INTERNAL_IMAGE $tag",
+        "workdir /root",
+    ]
+
+
+def infer_commands(pkg_root: Path) -> List[DockerCmdBlock]:
+    commands: List[DockerCmdBlock] = []
+
+    if (pkg_root / "system-requirements.txt").exists():
+        print("Adding system requirements installation phase")
+        commands.append(
+            DockerCmdBlock(
+                comment="install system requirements",
+                commands=[
+                    "copy system-requirements.txt /opt/latch/system-requirements.txt",
+                    "run apt-get update --yes && xargs apt-get install --yes </opt/latch/system-requirements.txt",
+                ],
+                order=DockerCmdBlockOrder.precopy,
+            )
+        )
+
+    if (pkg_root / "environment.R").exists():
+        print("Adding R + R package installation phase")
+        commands.append(
+            DockerCmdBlock(
+                comment="install R requirements",
+                commands=[
+                    dedent(
+                        r"""
+                        run apt-get update --yes && \
+                            apt-get install --yes software-properties-common && \
+                            add-apt-repository "deb http://cloud.r-project.org/bin/linux/debian buster-cran40/" && \
+                            apt-get install --yes r-base r-base-dev libxml2-dev libcurl4-openssl-dev libssl-dev wget
+                        """
+                    ).strip(),
+                    "copy environment.R /opt/latch/environment.R",
+                    "run Rscript /opt/latch/environment.R",
+                ],
+                order=DockerCmdBlockOrder.precopy,
+            )
+        )
+
+    conda_env_file_name = None
+    if (pkg_root / "environment.yml").exists():
```

Comment:
add to the docs that if both `environment.yaml` and `environment.yml` are present, `environment.yml` will take precedence and `environment.yaml` will be ignored